Automattic/pocket-casts-ios
# Bookmarks: Adds the Settings properties

| 🚨  Depends on: https://github.com/Automattic/pocket-casts-ios/pull/935 |
|:---:|

This adds the underlaying setting values for the headphone/bookmark settings. In normal fashion I also did some refactoring to hopefully make this a bit easier in the future. 

<details>
  <summary><h3>JSONCodable</h3></summary>
  
When storing enum values to the UserDefaults we have [previously](https://github.com/Automattic/pocket-casts-ios/blob/3601468f8c4d1af6a4f6ca20e4eec6815d2f9ded/podcasts/Settings.swift#L14) on using `Int` and storing the `rawValue` but this can end up causing issues if we want to change the order of enum values. The fix would be to manually define each value which, to me, defeats the purpose of magical enums. 

Instead of storing a raw value we can instead encode and decode the value using `JSONEncoder`/`JSONDecoder` but it also results in some extra boilerplate;

```swift
enum HelloWorld: Codable {
    case one, two, three
}

let value: HelloWorld = .two

// Save
let data = try? JSONEncoder().encode(value)
UserDefaults.standard.set(data, forKey: "hello_world")

// Get
if let data = UserDefaults.standard.data(forKey: "hello_world") {
    let value = try? JSONDecoder().decode(HelloWorld.self, from: data) 
    print(value)
}

print(value)

```

Instead we can now simplify it down to:

```swift
enum HelloWorld: JSONCodable {
    case one, two, three
}

let value: HelloWorld = .two
UserDefaults.standard.set(value, forKey: "hello_world")

let value = UserDefaults.standard.jsonObject(HelloWorld.self, forKey: "hello_world)
print(value)

```

If needed objects can override the default encoder/decoder to fit their needs. 

</details>

<details>
  <summary><h3>SettingValue</h3></summary>

Continuing with the trend of storing things in the UserDefaults, I've created a new struct named `SettingValue` that intends to reduce the boilerplate of adding a new setting and make it easier to understand the code. 

Right now, the items we store UserDefaults are defined in:
- The Key is stored in [`Constants.UserDefaults`](https://github.com/Automattic/pocket-casts-ios/blob/3601468f8c4d1af6a4f6ca20e4eec6815d2f9ded/podcasts/Constants.swift#L100)
- The default value is sometimes stored in:
    - [`AppDelegate+Defaults`](https://github.com/Automattic/pocket-casts-ios/blob/3601468f8c4d1af6a4f6ca20e4eec6815d2f9ded/podcasts/AppDelegate%2BDefaults.swift#L17)
    - As a fallback in the `Settings`
        - [Example 1](https://github.com/Automattic/pocket-casts-ios/blob/3601468f8c4d1af6a4f6ca20e4eec6815d2f9ded/podcasts/Settings.swift#L149)
        - [Example 2](https://github.com/Automattic/pocket-casts-ios/blob/3601468f8c4d1af6a4f6ca20e4eec6815d2f9ded/podcasts/Settings.swift#L170)
    - Other places 🤷‍♀️

This aims to unify this by moving the configuration, default value, setting and retrieving to one place. 

Using it is simple, and can be dropped into our current Constants setup. And it works with JSONCodable.

```swift
let setting = SettingValue("hello", defaultValue: "world")

print(setting.value()) -> defaults to "world"
setting.save("hello world")
print(setting.value()) -> prints "hello world"
```

I wrote more about the details and usage in the [documentation comment](https://github.com/Automattic/pocket-casts-ios/blob/8f1d7bd50a03d782f886e008f747e8b037ff475c/podcasts/Constants.swift#L314)
</details>

## To test
There's only code change in this so 👀 on the code and a 🟢 CI build should be enough. 

## Checklist
